File: src/main/python/AoC2023_21.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 21
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.common import log
from aoc.grid import CharGrid

Input = CharGrid
Output1 = int
Output2 = int


TEST = """\
...........
.....###.#.
.###.##..#.
..#.#...#..
....#.#....
.##..S####.
.##..#...#.
.......##..
.##.#.####.
.##..##.##.
...........
"""

DIRS = {(0, 1), (0, -1), (1, 0), (-1, 0)}
STEPS = 26_501_365


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return CharGrid.from_strings([_ for _ in input_data])

    def solve(self, grid: CharGrid, steps: list[int]) -> list[int]:
        w = grid.get_width()
        start = (w // 2, w // 2)
        plots = set[tuple[int, int]]([start])
        ans = []
        for i in range(1, max(steps) + 1):
            new_plots = set[tuple[int, int]]()
            for r, c in plots:
                for dr, dc in DIRS:
                    rr, cc = r + dr, c + dc
                    wr, wc = rr % w, cc % w
                    if (
                        0 <= wr < w
                        and 0 <= wc < w
                        and grid.values[wr][wc] != "#"
                    ):
                        new_plots.add((rr, cc))
            plots = new_plots
            if i in steps:
                ans.append(len(plots))
        return ans

    def part_1(self, grid: Input) -> Output1:
        return self.solve(grid, [64])[0]

    def part_2(self, grid: Input) -> Output2:
        w = grid.get_width()
        modulo = STEPS % w
        x = STEPS // w
        # https://old.reddit.com/r/adventofcode/comments/18nni6t/2023_21_part_2_optimization_hint/
        steps = [w - modulo - 2] + [i * w + modulo for i in range(2)]
        log(steps)
        values = self.solve(grid, steps)
        log(values)
        a = (values[2] + values[0]) // 2 - values[1]
        b = (values[2] - values[0]) // 2
        c = values[1]
        log((f"{a=}", f"{b=}", f"{c=}"))
        return a * x * x + b * x + c

    @aoc_samples(())
    def samples(self) -> None:
        pass


solution = Solution(2023, 21)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_21.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_21 extends SolutionBase<CharGrid, Long, Long> {
    
    private static final int STEPS = 26_501_365;
    
    private AoC2023_21(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_21 create() {
        return new AoC2023_21(false);
    }
    
    public static AoC2023_21 createDebug() {
        return new AoC2023_21(true);
    }
    
    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return CharGrid.from(inputs);
    }

    @Override
    public Long solvePart1(final CharGrid grid) {
        return (long) this.solve(grid, List.of(64)).get(0);
    }
    
    @Override
    public Long solvePart2(final CharGrid grid) {
        final int w = grid.getWidth();
        final int mod = STEPS % w;
        final int x = STEPS / w;
        // https://old.reddit.com/r/adventofcode/comments/18nni6t/2023_21_part_2_optimization_hint/
        final List<Integer> steps = new ArrayList<>();
        steps.add(w - mod - 2);
        range(2).intStream().map(i -> i * w + mod).forEach(steps::add);
        log(steps);
        final List<Integer> values = this.solve(grid, steps);
        log(values);
        final long a = (values.get(2) + values.get(0)) / 2 - values.get(1);
        final long b = (values.get(2) - values.get(0)) / 2;
        final long c = values.get(1);
        log("a: %d, b: %d, c: %d".formatted(a, b, c));
        return a * x * x + b * x + c;
    }
    
    private record Plot(long r, long c) {}
    
    private List<Integer> solve(final CharGrid grid, final List<Integer> steps) {
        final int w = grid.getWidth();
        Set<Plot> plots = new HashSet<>();
        plots.add(new Plot(w / 2, w / 2));
        final List<Integer> ans = new ArrayList<>();
        final int max = steps.stream().mapToInt(Integer::intValue).max().getAsInt();
        for (int i = 1; i <= max; i++) {
            final Set<Plot> newPlots = new HashSet<>();
            for (final Plot p : plots) {
                for (final Direction d : Direction.CAPITAL) {
                    final long rr = p.r() + d.getY();
                    final long cc = p.c() + d.getX();
                    final int wr = Math.floorMod(rr, w);
                    final int wc = Math.floorMod(cc, w);
                    if (0 <= wr && wr < w && 0 <= wc && wc < w
                            && grid.getValue(Cell.at(wr, wc)) != '#') {
                        newPlots.add(new Plot(rr, cc));
                    }
                }
            }
            plots = newPlots;
            if (steps.contains(i)) {
                ans.add(plots.size());
            }
        }
        return ans;
    }
    
    @Override
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_21.create().run();
    }
}